File: solution/java/lengthOfLongestSubstring.java

import java.util.*;

public class lengthOfLongestSubstring {
    public static void main(String[] args) {

    }

    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> prevIndex = new HashMap<>();
        int res = 0;

        for (int lo = 0, hi = 0; hi < s.length(); hi++) {
            int i = prevIndex.getOrDefault(s.charAt(hi), -1);

            while(lo <= i) {
                lo++;
            }
            prevIndex.put(s.charAt(hi), hi);
            res = Math.max(res, hi - lo + 1);
        }

        return res;
    }

}

File: solution/js/lengthOfLongestSubstring.js
/**
 * @param {string} s
 * @return {number}
 */

var lengthOfLongestSubstring = function (s) {
  const prevIndex = new Map();
  let res = 0;

  for (let lo = 0, hi = 0; hi < s.length; hi++) {
    let i = prevIndex.get(s[hi]) ?? -1;

    while (lo <= i) {
      lo++;
    }
    res = Math.max(res, hi - lo + 1);
    prevIndex.set(s[hi], hi);
  }
  return res;
};

console.log(lengthOfLongestSubstring("abcabcbb"));
